src/subject_name/dns_name.rs

```diff
@@ -130,24 +123,145 @@ impl<'a> DnsNameRef<'a> {
     }
 }
 
-/// Requires the `alloc` feature.
-#[cfg(feature = "alloc")]
 impl core::fmt::Debug for DnsNameRef<'_> {
     fn fmt(&self, f: &mut core::fmt::Formatter) -> Result<(), core::fmt::Error> {
-        let lowercase = self.clone().to_owned();
-        f.debug_tuple("DnsNameRef").field(&lowercase.0).finish()
+        f.write_str("DnsNameRef(\"")?;
+
+        // Convert each byte of the underlying ASCII string to a `char` and
+        // downcase it prior to formatting it. We avoid self.clone().to_owned()
+        // since it requires allocation.
+        for &ch in self.0 {
+            f.write_char(char::from(ch).to_ascii_lowercase())?;
+        }
+
+        f.write_str("\")")
     }
 }
 
-#[cfg(not(feature = "alloc"))]
-impl core::fmt::Debug for DnsNameRef<'_> {
+impl<'a> From<DnsNameRef<'a>> for &'a str {
+    fn from(DnsNameRef(d): DnsNameRef<'a>) -> Self {
+        // The unwrap won't fail because DnsNameRefs are guaranteed to be ASCII
+        // and ASCII is a subset of UTF-8.
+        core::str::from_utf8(d).unwrap()
+    }
+}
+
+/// A DNS Name suitable for use in the TLS Server Name Indication (SNI)
+/// extension and/or for use as the reference hostname for which to verify a
+/// certificate.
+pub enum GeneralDnsNameRef<'name> {
+    /// a valid DNS name
+    DnsName(DnsNameRef<'name>),
+    /// a DNS name containing a wildcard
+    Wildcard(WildcardDnsNameRef<'name>),
+}
+
+impl<'a> From<GeneralDnsNameRef<'a>> for &'a str {
+    fn from(d: GeneralDnsNameRef<'a>) -> Self {
+        match d {
+            GeneralDnsNameRef::DnsName(name) => name.into(),
+            GeneralDnsNameRef::Wildcard(name) => name.into(),
+        }
+    }
+}
+
+/// A reference to a DNS Name suitable for use in the TLS Server Name Indication
+/// (SNI) extension and/or for use as the reference hostname for which to verify
+/// a certificate. Compared to `DnsName`, this one will store domain names containing
+/// a wildcard.
+///
+/// A `WildcardDnsName` is guaranteed to be syntactically valid. The validity rules are
+/// specified in [RFC 5280 Section 7.2], except that underscores are also
+/// allowed, and following [RFC 6125].
+///
+/// `WildcardDnsName` stores a copy of the input it was constructed from in a `String`
+/// and so it is only available when the `alloc` default feature is enabled.
+///
+/// [RFC 5280 Section 7.2]: https://tools.ietf.org/html/rfc5280#section-7.2
+/// [RFC 6125]: https://tools.ietf.org/html/rfc6125
+#[cfg(feature = "alloc")]
+#[derive(Clone, Debug, Eq, PartialEq, Hash)]
+pub struct WildcardDnsName(String);
+
+#[cfg(feature = "alloc")]
+impl WildcardDnsName {
+    /// Returns a `WildcardDnsNameRef` that refers to this `WildcardDnsName`.
+    pub fn as_ref(&self) -> WildcardDnsNameRef {
+        WildcardDnsNameRef(self.0.as_bytes())
+    }
+}
+
+#[cfg(feature = "alloc")]
+impl AsRef<str> for WildcardDnsName {
```

Comment:
There's an `AsRef<str>` for `WildcardDnsName` but not for `WildcardDnsNameRef`. I think we probably want it for the latter, too?

Reply:
:thinking: I notice `DnsName` has `AsRef<str>`, but `DnsNameRef` only has `AsRef<[u8]>`.

I think it makes sense to add `AsRef<[u8]>` to `WildcardDnsNameRef`, but I'm not sure about `AsRef<str>`. If we add that to `WildcardDnsNameRef` with a `core::str::from_utf8(self.0).unwrap()` should it also be added to `DnsNameRef` to maintain symmetry?

Reply:
I think `AsRef<str>` makes sense -- we know the contents are valid ASCII and `AsRef<str>` is strictly more capable than `AsRef<[u8]>` (for one thing, you can trivially call `as_bytes()` on `&str` to get `&[u8]`). So IMO this merely makes the guarantee that the contents of `WildcardDnsNameRef` are valid UTF-8, which seems like something we'll want to uphold anyway.

As for adding `AsRef<str>` for `DnsNameRef`, I think this could be a bit of a breaking API change because type inference for `.as_ref()` calls will stop doing the right thing automatically. So, while I think it would be a good thing to do, I don't think we want to do it now. Maybe file an issue or PR to do this on the next API-incompatible change? (We can probably remove the `AsRef<[u8]>` in that case, for the reasons mentioned above.)

Reply:
https://github.com/rustls/webpki/issues/45